### PR TITLE
Improve `poll()`/`select()`

### PR DESCRIPTION
**Pull request by jwt27**
Now `poll()` no longer calls `select_s()`, so the whole `fd_set` dance is avoided.  Should be significantly faster, and saves a good 30KB (!) of stack space.

The `POLLNVAL`/`POLLHUP`/`POLLERR` flags are now also used, in accordance with the Linux manpage.

### CODE REVIEW

**Comment by gvanem** [comment]
The CI completed fine now. This looks effin awesome.
So `poll.c` now gets removed since it needs internals in `select.c` or what? Seems OK by me.

**Comment by jwt27** [comment]
> The CI completed fine now. This looks effin awesome.

Thanks!

> So `poll.c` now gets removed since it needs internals in `select.c` or what? Seems OK by me.

Yes, I moved `poll()` to `select.c` since there is a lot of shared code.

I did some benchmarking by connecting to an invalid TCP port, and then counting the number of calls to `poll()` while waiting for the 30 second timeout to expire.  Like so:
```c
  int i = 0;
  struct pollfd pfd;
  do
  {
    pfd.fd = fd;
    pfd.events = POLLOUT;
    pfd.revents = 0;
    if (poll(&pfd, 1, 0) < 0)
    {
      perror("poll()");
      goto fail;
    }
    ++i;
  } while (pfd.revents == 0);
```

The result is somewhat surprising:
```
- old poll() via select_s():
42498
33644
19915
76479
75630
63016
65807
60339
47633
50679

- new poll():
108432
104442
107909
107820
108115
108166
108024
106090
108581
108567
```

So the `select_s()` code is highly variable for some reason.  Might be worth looking into why that is.
The good news is, the new `poll()` is more consistent, and up to 445% faster :)

**Comment by gvanem** [comment]
> while waiting for the 30 second timeout to expire

Those numbers are `i` after the loops has finished? I assume those 30 sec are
set in some test-program of your, which protocol? And you're sure 30 sec elapsed in both tests?

BTW, from Linux' `man select`:
```
 On Linux, select() modifies timeout to reflect the amount of time not slept;
```
(or remaining) I implemented that once. Hope this still holds with this PR.

**Comment by jwt27** [comment]
> > while waiting for the 30 second timeout to expire
> 
> Those numbers are `i` after the loops has finished? I assume those 30 sec are set in some test-program of your, which protocol? And you're sure 30 sec elapsed in both tests?

It just waits for `socket->nb_timer` to expire (from [here](https://github.com/gvanem/Watt-32/blob/cb146e4dc2948b741a801db2328eed07d00ea8df/src/connect.c#L372)).  Accuracy depends on the rdtsc code in `gettimeofday2()`, but I thought 10 runs should produce a usable average.

> BTW, from Linux' `man select`:
> 
> ```
>  On Linux, select() modifies timeout to reflect the amount of time not slept;
> ```
> 
> (or remaining) I implemented that once. Hope this still holds with this PR.

Should still work, I didn't touch it.

File: src/select.c [hunk 1]
```diff
@@ -263,13 +258,13 @@ int W32_CALL select_s (int nfds, fd_set *readfds, fd_set *writefds,
...
       SOCK_DBUG_FLUSH();
 
     } /* end of for loop; all sockets checked at least once */
 
+    /* Safe to run other "tasks" now.
+     */
+    _sock_crit_stop();
 
     /* WATT_YIELD() sometimes hangs for approx 250msec under Win-XP.
      * Don't yield for "small" timeouts.
```
Comment (jwt27) [inline src/select.c:270]: BTW, there is this bit in `select_s()`.  Do we want the same logic for `poll()` too?  Or maybe remove it from `select_s()` (don't know how relevant WinXP still is).

Problem here is, if someone did `sock_yield(NULL, my_yield)`, that will never be called either.

Comment (gvanem) [reply src/select.c:270]: It should yield if `ret_count == 0` in the hope that some events happen in the mean time.

>  (don't know how relevant WinXP still is).

Not sure if it's an issue with Win-10. Some test is needed to verify.

> Problem here is, if someone did sock_yield(NULL, my_yield), that will never be called either.

That `my_yield` won't be called? I think it should be called indirectly as:
```c
      if (socket->tcp_sock->usr_yield)
         (*socket->tcp_sock->usr_yield) ();
```
in lots of places.

**Review by gvanem — COMMENTED** [review]
I hit *Review* when I mean to *Comment*.

File: src/select.c (hunk 1 above)
Comment (jwt27) [reply src/select.c:270]: > It should yield if `ret_count == 0` in the hope that some events happen in the mean time.

It does, but `select_s()` skips that if the timeout is less than 500ms.  Seems to be the only place where that is done.

Looking at `WATT_YIELD` again, I see now it expands to `os_yield()` (eg. `int 2f, ax=1680`) directly.  Huh.  I thought it called `system_yield`.  Maybe it should?

Comment (gvanem) [reply src/select.c:270]: >  Maybe it should?

I don't see a problem here. What is it?

Comment (jwt27) [reply src/select.c:270]: There are two, basically:
* `WATT_YIELD()` doesn't call the user-provided yield function
* `select_s()` doesnt call `WATT_YIELD()` if timeout is less than 500ms

I think if the user has set a custom global yield function, that should take precedence over the OS one.  
Then for WinXP maybe some other workaround can be found (eg. only yield once every X ms).

And again, thanks for merging!

Comment (jwt27) [reply src/select.c:270]: Do you happen to remember if this comment in `select.c` applies to a native winxp program, or a dos program running under winxp?  I'm guessing it's only the latter?

Comment (gvanem) [reply src/select.c:270]: >  I'm guessing it's only the latter?

AFAICR, yes. I even ran djgpp/Watt-32 programs under Win-95 (or was it Win-98?). 
I think Win-XP made it worse for djgpp; hence  that comment. But again, I'm rusty on these details after such a long time.
And now after 2 pints of Tuborgs.

Comment (jwt27) [reply src/select.c:270]: I can sort-of replicate this on windows 2000, but I'm not sure if it's the same thing.  Using a djgpp test program, doing the following in a loop:

```c
t0 = set_timeout(0);
__dpmi_yield();
t = set_timeout(0) - t0;
```

I'm regularly seeing `t` values around 1000ms.  But it doesn't feel like it's stalling for a whole second.  And it often jumps backwards too (`t < -900`)...  Maybe the rdtsc code isn't very accurate on Windows?  The high `t` values also completely disappear after using `init_timer_isr()` (which disables rdtsc stuff).

File: src/select.c [hunk 2]
```diff
@@ -378,6 +375,103 @@ int W32_CALL select (int nfds, fd_set *read_fds, fd_set *write_fds,
...
+  _sock_sig_setup();
+
+  /*
+   * Loop until specified timeout expires or event(s) satisfied.
+   */
+  while (1)
+  {
+    /* Not safe to run sock_daemon() (or other "tasks") now
+     */
+    _sock_crit_start();
```
Comment (jwt27) [inline src/select.c:413]: Looking at this again, I don't think this is necessary at all.  If the socket gets deleted it just returns `POLLNVAL`.

In fact, `_sock_crit_stop()` will now run the daemon on every loop so this only hurts performance.

**Pull request merged**